### ISSUE TITLE
cinterop for Google's Gumbo Parser

### ISSUE THREAD

**Issue by tech4242** [JetBrains/kotlin-native#1417]
I completed the [official Kotlin/Native tutorials](https://kotlinlang.org/docs/tutorials/native/interop-with-c.html) a short while ago and I really liked the cinterop explanation. I also implemented the curl example with a Makefile in [my own repo](https://github.com/tech4242/kotlin-native-crawler). 

My question: How would I go about "cinterop-ing" [Google's Gumbo Parser](https://github.com/google/gumbo-parser) (written in C99) to extend the functionality of your example. I want to understand how to go through the motions with bigger external libraries - if at all possible. And why not extend your official example for this.

**Comment by olonho** [comment]
What is the difference between Gumbo Parser and other libraries? Do you have troubles running interop on it?

**Comment by tech4242** [comment]
Yes, currently on [this](https://github.com/tech4242/kotlin-native-crawler/tree/cinterop-gumbo) branch. I installed the parser using [Google's official guide](https://github.com/google/gumbo-parser):

```
$ ./autogen.sh
$ ./configure
$ make
$ sudo make install
```
Then I added the src to my project, added the new gumbo.def file & added it to my makefile. I get the following from make:

```
mkdir -p build
mkdir -p build/c_interop
mkdir -p build/kotlin
cinterop -def src/c_interop/libcurl.def -o build/c_interop/libcurl
cinterop -def src/c_interop/gumbo.def -o build/c_interop/gumbo
Exception in thread "main" java.lang.Error: /var/folders/8q/03534j193yxbhkxdmc1z0r1w0000gn/T/tmp333700181360103801.c:1:10: fatal error: 'gumbo/gumbo.h' file not found
	at org.jetbrains.kotlin.native.interop.indexer.UtilsKt.ensureNoCompileErrors(Utils.kt:137)
	at org.jetbrains.kotlin.native.interop.indexer.IndexerKt.indexDeclarations(Indexer.kt:898)
	at org.jetbrains.kotlin.native.interop.indexer.IndexerKt.buildNativeIndexImpl(Indexer.kt:888)
	at org.jetbrains.kotlin.native.interop.indexer.NativeIndexKt.buildNativeIndex(NativeIndex.kt:56)
	at org.jetbrains.kotlin.native.interop.gen.jvm.MainKt.processCLib(main.kt:307)
	at org.jetbrains.kotlin.native.interop.gen.jvm.MainKt.interop(main.kt:38)
	at org.jetbrains.kotlin.cli.utilities.InteropCompilerKt.invokeInterop(InteropCompiler.kt:100)
	at org.jetbrains.kotlin.cli.utilities.MainKt.main(main.kt:29)
make: *** [interop] Error 1
```

And that sort of makes sense I guess. I am not sure how the gumbo.h should be imported within the def file. So I guess right now I am struggling with the path definition for the header in gumbo.def. I have added gunbo's src to my own project src in a folder next to my def files.

**Comment by olonho** [comment]
Likely you need something like
```
compilerOpts.linux = -I/usr/local/include
compilerOpts.osx = -I/usr/local/include -I/usr/include/x86_64-linux-gnu
```
or wherever your header files are installed, so that interop tool could find them.

**Comment by olonho** [comment]
Does it fixes your problem?

**Comment by tech4242** [comment]
Sadly, I am not quite there yet. I also looked at this issue https://github.com/JetBrains/kotlin-native/issues/1032 and tried the def file with:
```
headers = gumbo.h
compilerOpts = -I/src/c/gumbo
```

where `src/c/gumbo` is the local path from the project folder. I think I just need a crash course on how path referencing works in def files. That is pretty much the reason I tried to expand the official tutorial with a much larger external C library - unless you are really into cinterop, you wouldn't know how this works. I want to have my repo as a reference for external C libs with Konan.
